### PR TITLE
Added one more Data Structure - Arrays

### PR DESCRIPTION
**Pull request by Swastyy**
In the Data Structure folder I have added the brief explanation of arrays.

### CODE REVIEW

File: en/Data Structures/Arrays/array.md
Comment (Panquesito7) [inline en/Data Structures/Arrays/array.md:3]:
```suggestion
Arrays are the most basic data structures used in programming. Arrays are data structures that contain a group of elements. Typically these elements are all of the same data type, such as `char` or `int`.\
```

File: en/Data Structures/Arrays/array.md
Comment (Panquesito7) [inline en/Data Structures/Arrays/array.md:11]:
```suggestion
`type` is the datatype of the array elements
```

File: en/Data Structures/Arrays/array.md
Comment (Panquesito7) [inline en/Data Structures/Arrays/array.md:18]:
```suggestion
```

File: en/Data Structures/Arrays/array.md
Comment (Panquesito7) [inline en/Data Structures/Arrays/array.md:22]:
```suggestion
- [Introduction to arrays](https://www.geeksforgeeks.org/introduction-to-arrays/)
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
👍 Good work!

File: en/Data Structures/Arrays/array.md
Comment (Panquesito7) [inline en/Data Structures/Arrays/array.md:4]:
```suggestion
All arrays consist of contiguous memory locations. We can access the elements with their index. The lowest address corresponds to the first element, and the highest address to the last element.

```

**Review by Panquesito7 — DISMISSED** [review]
LGTM. Thank you for your contribution! 👍 🎉
@appgurueu, would you like to review this PR? Thanks. 🙂

**Comment by appgurueu** [comment]
Oddly programming language (C) specific. I'd want something more abstract: Arrays are used to represent ordered sets of elements in memory accessible through indices in O(1) constant time (usually through pointer arithmetics, as you noted, but that's implementation specific). Array size is fixed, other than, say, linked list size.

**Comment by Swastyy** [comment]
> Oddly programming language (C) specific. I'd want something more abstract: Arrays are used to represent ordered sets of elements in memory accessible through indices in O(1) constant time (usually through pointer arithmetics, as you noted, but that's implementation specific). Array size is fixed, other than, say, linked list size.

Ohh then we can use vectors instead more specifically vector arrays. I mean there will not be size restriction and accessing elements using index will still be there. Moreover arrays (normal and vector both) are very commonly used by everyone. Let me know your thoughts

File: en/Data Structures/Arrays/array.md
Comment (raklaptudirm) [inline en/Data Structures/Arrays/array.md:3]:
```suggestion
Arrays are the most basic data structures used in programming. They contain a group of a fixed number of elements which occupy the same space in memory, i.e., the memory occupied is known at compile time. Thus arrays are stored in the stack and it's values can be saved and retrieved very quickly, in `O(1)` time. Languages like `Rust` which allows for `enums` in Arrays, each element occupies the maximum space which could be required by it.
```

File: en/Data Structures/Arrays/array.md
Comment (raklaptudirm) [inline en/Data Structures/Arrays/array.md:10]: The declaration, initialization works for only certain c-like languages. Would consider removing them.

File: en/Data Structures/Arrays/array.md
Comment (raklaptudirm) [inline en/Data Structures/Arrays/array.md:3]:
```suggestion
Arrays are the most basic data structures used in programming. They contain a group of a fixed number of elements which occupy the same space in memory, i.e., the memory occupied is known at compile time. Thus arrays are stored in the stack and it's values can be saved and retrieved very quickly, in `O(1)` time. Languages like `Rust` which allows for `enums` in Arrays, each element occupies the maximum space which could be required by it.
```

**Review by raklaptudirm — CHANGES_REQUESTED** [review]
Some suggestions to make the concept of arrays more clear and to make them more general, i.e., not oriented to a single language.

File: en/Data Structures/Arrays/array.md
Comment (raklaptudirm) [reply en/Data Structures/Arrays/array.md]: Also do this for the Array initialization and element access. They are not important nor unique to arrays.

Comment (Swastyy) [reply en/Data Structures/Arrays/array.md]: Done👍🏻

**Review by raklaptudirm — DISMISSED** [review]

File: en/Data Structures/Arrays/array.md
Comment (Panquesito7) [inline en/Data Structures/Arrays/array.md:3]:
```suggestion
Arrays are the most basic data structures used in programming. They contain a group of a fixed number of elements that occupy the same space in memory. I.e., the memory occupied is known at compile time. Thus arrays are stored in the stack, and their values can be saved and retrieved very quickly in `O(1)` time. Languages like `Rust` which allows `enums` in Arrays, each element occupies the maximum space which could be required by it.
```

File: en/Data Structures/Arrays/array.md
Comment (Panquesito7) [inline en/Data Structures/Arrays/array.md:12]:
```suggestion
- [YouTube URL tutorial video](https://youtu.be/NptnmWvkbTw)
```

File: en/Data Structures/Arrays/array.md
Comment (Panquesito7) [inline en/Data Structures/Arrays/array.md:13]:
```suggestion
- [YouTube video for understanding Arrays in C++](https://youtu.be/ibeGtDEQGz0)
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Almost there! 😄

**Review by raklaptudirm — DISMISSED** [review]

File: en/Data Structures/Arrays/array.md
Comment (appgurueu) [inline en/Data Structures/Arrays/array.md:3]: The memory occupied is not necessarily known at compile time. This depends on your array definition; Go calls arrays with fixed length at compile time arrays and arrays with "fixed" length at runtime slices.

**Review by Panquesito7 — APPROVED** [review]
LGTM. Although, I think you should check out @appgurueu's review before merging.

**Comment by Swastyy** [comment]
> LGTM. Although, I think you should check out @appgurueu's review before merging.

Yep👍🏻

File: en/Data Structures/Arrays/array.md
Comment (Swastyy) [reply en/Data Structures/Arrays/array.md:3]: Considering that we are defining array size while initialization then memory size will be fixed. I think by stack @raklaptudirm mean some kind of pile. And the rust is just to take as an example of variant that also exist ig. 😕

**Review by vbrazo — APPROVED** [review]

File: en/Data Structures/Arrays/array.md
Comment (appgurueu) [reply en/Data Structures/Arrays/array.md:3]: > Considering that we are defining array size while initialization then memory size will be fixed. I think by stack @raklaptudirm mean some kind of pile. And the rust is just to take as an example of variant that also exist ig. confused

Memory size will usually be fixed, yes. But note that stack != heap. Java stores objects, including arrays, on the heap.

Comment (Swastyy) [reply en/Data Structures/Arrays/array.md:3]: Yeah, stack is not used generally for dynamic size variables.

**Review by Panquesito7 — APPROVED** [review]
I'll merge in a few days if there are no objections.

**Comment by appgurueu** [comment]
> I'll merge in a few days if there are no objections.

I do object. As said:

1. Remove the comment on arrays being stored "on the stack" (what is this supposed to mean anyways? implementation specific)
2. Remove the comment regarding memory layout in Rust

**Comment by Swastyy** [comment]
> > I'll merge in a few days if there are no objections.
> 
> I do object. As said:
> 
> 1. Remove the comment on arrays being stored "on the stack" (what is this supposed to mean anyways? implementation specific)
> 2. Remove the comment regarding memory layout in Rust

Alright. Using stack is anyways very bad.

**Review by raklaptudirm — DISMISSED** [review]

File: en/Data Structures/Arrays/array.md
Comment (Panquesito7) [inline en/Data Structures/Arrays/array.md:8]:
```suggestion
- [Introduction to arrays - GeeksForGeeks](https://www.geeksforgeeks.org/introduction-to-arrays/)
```

**Review by Panquesito7 — APPROVED** [review]

**Review by raklaptudirm — APPROVED** [review]

**Comment by Panquesito7** [comment]
@appgurueu, would you like to review? Thanks.

**Comment by appgurueu** [comment]
> Alright. Using stack is anyways very bad.

No it's not? A linked stack has constant time pushing and popping.

**Comment by Swastyy** [comment]
> > Alright. Using stack is anyways very bad.
> 
> No it's not? A linked stack has constant time pushing and popping.
> 
> Still, the compile time remark depends on your terminology. In Golang terms, this is correct; arrays whose size is known at runtime are called slices. Array size is fixed at allocation.

I think we shouldn't take different languages as examples since every data structure is not same in every language in terms of declaration or working. We are creating this array file assuming that the person seeing will use it where it can be applied in the way.

**Comment by appgurueu** [comment]
> > > Alright. Using stack is anyways very bad.
> > 
> > 
> > No it's not? A linked stack has constant time pushing and popping.
> > Still, the compile time remark depends on your terminology. In Golang terms, this is correct; arrays whose size is known at runtime are called slices. Array size is fixed at allocation.
> 
> I think we shouldn't take different languages as examples since every data structure is not same in every language in terms of declaration or working. We are creating this array file assuming that the person seeing will use it where it can be applied in the way.

This isn't about languages, it is about terminology. Usually, the term "array" is **not** used to **only** refer to arrays whose size is fixed at **compile time**. That's a Golang thing. Array size is fixed at allocation. In C for example you can do `int n = runtime value; int array[n];`.

**Comment by Swastyy** [comment]
> This isn't about languages, it is about terminology. Usually, the term "array" is **not** used to **only** refer to arrays whose size is fixed at **compile time**. That's a Golang thing. Array size is fixed at allocation. In C for example you can do `int n = runtime value; int array[n];`.

`The actual array will be allocated at runtime, using the size defined at compile time`
See for reference https://stackoverflow.com/questions/26175859/is-the-size-of-an-array-determined-at-compile-time

**Comment by appgurueu** [comment]
> > This isn't about languages, it is about terminology. Usually, the term "array" is **not** used to **only** refer to arrays whose size is fixed at **compile time**. That's a Golang thing. Array size is fixed at allocation. In C for example you can do `int n = runtime value; int array[n];`.
> 
> `The actual array will be allocated at runtime, using the size defined at compile time`
> See for reference https://stackoverflow.com/questions/26175859/is-the-size-of-an-array-determined-at-compile-time

See https://stackoverflow.com/questions/2672085/static-array-vs-dynamic-array-in-c. You're calling only static arrays arrays, but dynamic arrays are arrays too. It is important to include this distinction.

**Comment by Swastyy** [comment]
You're calling only static arrays arrays, but dynamic arrays are arrays too. It is important to include this distinction.

Correct. Review it once.

**Review by Panquesito7 — APPROVED** [review]

**Review by vbrazo — APPROVED** [review]

**Pull request merged**
